### PR TITLE
Handle export exceptions

### PR DESCRIPTION
**Pull request by ruuushhh**

### CODE REVIEW

File: apps/quickbooks_online/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+from apps.tasks.models import TaskLog, Error
+from apps.fyle.models import ExpenseGroup
+from qbosdk.exceptions import WrongParamsError, InvalidTokenError
+
+
+logger = logging.getLogger(__name__)
+logger.level = logging.INFO
+
+
+def handle_quickbooks_error(exception, expense_group: ExpenseGroup, task_log: TaskLog, export_type: str):
```
Comment (ashwin1111) [inline apps/quickbooks_online/exceptions.py:17]: hope nothing is changed in this function, pls reply incase if there are any changes

File: apps/quickbooks_online/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+def handle_export_exceptions(task_name):
+    def decorator(func):
+        def new_fn(expense_group: ExpenseGroup, task_log_id: int, *args):
+            error = {
+                'task': task_name,
+                'workspace_id': expense_group.workspace_id,
+                'alert': False,
+                'message': None,
+                'response': None
+            }
```
Comment (ashwin1111) [inline apps/quickbooks_online/exceptions.py:71]: I don't think this is needed at all

File: apps/quickbooks_online/tasks.py [hunk 3]
```diff
@@ -808,6 +693,7 @@ def schedule_qbo_expense_creation(workspace_id: int, expense_group_ids: List[str
             chain.run()
 
 
+@handle_export_exceptions(task_name="Create QBO Expense")
```
Comment (ashwin1111) [inline apps/quickbooks_online/tasks.py:696]: I hope you would've removed try and moved all logic 1 block outside, let me know if you've changed any other things

File: apps/quickbooks_online/exceptions.py (hunk 1 above)
Comment (ruuushhh) [reply apps/quickbooks_online/exceptions.py:17]: I just moved that function in exceptions file everything else is same

File: apps/quickbooks_online/exceptions.py (hunk 2 above)
Comment (ruuushhh) [reply apps/quickbooks_online/exceptions.py:71]: We will need this in default exception right?
`except Exception:
                error = traceback.format_exc()
                task_log.detail = {
                    'error': error
                }
                task_log.status = 'FATAL'
                task_log.save()
                logger.error('Something unexpected happened workspace_id: %s %s', 
                             task_log.workspace_id, task_log.detail)`

File: apps/quickbooks_online/tasks.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
 from .models import Bill, BillLineitem, Cheque, ChequeLineitem, CreditCardPurchase, CreditCardPurchaseLineitem, \
     JournalEntry, JournalEntryLineitem, BillPayment, BillPaymentLineitem, QBOExpense, QBOExpenseLineitem
 from .utils import QBOConnector
+from .exceptions import handle_export_exceptions, handle_quickbooks_error
```
[the remark this answers is not in the document]
Comment (ruuushhh) [reply apps/quickbooks_online/tasks.py:27]: for `create_bill_payment` function as it takes `workspace_id` as default argument so we won't be able to use decorator for that

File: apps/quickbooks_online/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                        'is_resolved': False
+                    }
+                )
+
+    task_log.status = 'FAILED'
+    task_log.detail = None
+    task_log.quickbooks_errors = errors
+    task_log.save()
+
+def handle_export_exceptions(task_name):
```
[the remark this answers is not in the document]
Comment (ruuushhh) [reply apps/quickbooks_online/exceptions.py:62]: if we are using error then we might need this right?
https://github.com/fylein/fyle-qbo-api/pull/441#discussion_r1219640153

File: apps/quickbooks_online/tasks.py (hunk 3 above)
Comment (ruuushhh) [reply apps/quickbooks_online/tasks.py:696]: yes will check once again

File: apps/quickbooks_online/exceptions.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashwin1111) [reply apps/quickbooks_online/exceptions.py:62]: replied

File: apps/quickbooks_online/tasks.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashwin1111) [reply apps/quickbooks_online/tasks.py:27]: Screenshot 2023-06-08 at 4 15 08 PM
duplicate 23-51, move it to a separate function. Create 1 more decorator for handling bill payment exception. Expect args according to what you need, then re use the same function that we separated out above.

File: apps/quickbooks_online/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                        'is_resolved': False
+                    }
+                )
+
+    task_log.status = 'FAILED'
+    task_log.detail = None
+    task_log.quickbooks_errors = errors
+    task_log.save()
+
+def handle_export_exceptions():
```
Comment (ashwin1111) [inline apps/quickbooks_online/exceptions.py:62]: handle bill payment too

File: apps/quickbooks_online/exceptions.py [hunk 7]
```diff
@@ -59,12 +59,18 @@ def handle_quickbooks_error(exception, expense_group: ExpenseGroup, task_log: Ta
...
-def handle_export_exceptions():
+def handle_export_exceptions(bill_payment=False):
     def decorator(func):
-        def new_fn(expense_group: ExpenseGroup, task_log_id: int, *args):
-            task_log = TaskLog.objects.get(id=task_log_id)
+        def new_fn(*args):
+            if not bill_payment:
+                task_log_id = args[1]
+                task_log = TaskLog.objects.get(id=task_log_id)
+                expense_group = args[0]
```
Comment (ashwin1111) [inline apps/quickbooks_online/exceptions.py:68]: add this in line 66 for better readability

File: apps/quickbooks_online/exceptions.py [hunk 8]
```diff
@@ -59,12 +59,18 @@ def handle_quickbooks_error(exception, expense_group: ExpenseGroup, task_log: Ta
     task_log.quickbooks_errors = errors
     task_log.save()
 
-def handle_export_exceptions():
+def handle_export_exceptions(bill_payment=False):
     def decorator(func):
-        def new_fn(expense_group: ExpenseGroup, task_log_id: int, *args):
-            task_log = TaskLog.objects.get(id=task_log_id)
+        def new_fn(*args):
+            if not bill_payment:
```
Comment (ashwin1111) [inline apps/quickbooks_online/exceptions.py:65]: can we handle `last_export` as well?

File: apps/quickbooks_online/exceptions.py [hunk 9]
```diff
@@ -59,12 +59,18 @@ def handle_quickbooks_error(exception, expense_group: ExpenseGroup, task_log: Ta
...
+                task_log_id = args[1]
+                task_log = TaskLog.objects.get(id=task_log_id)
+                expense_group = args[0]
+            else:
+                expense_group = args[0].expense_group
```
Comment (ashwin1111) [inline apps/quickbooks_online/exceptions.py:70]: `def process_bill_payments(bill: Bill, workspace_id: int, task_log: TaskLog):    `
1st argument expects bill. 2nd arg workspace id is not added in else case. Check 3rd arg as well

File: apps/quickbooks_online/tasks.py [hunk 10]
```diff
@@ -945,6 +945,40 @@ def check_expenses_reimbursement_status(expenses):
 
     return all_expenses_paid
 
+@handle_export_exceptions(bill_payment=True)
```
Comment (ashwin1111) [inline apps/quickbooks_online/tasks.py:948]: let's name the decorator handle_qbo_exceptions if we are able to bring all handing under this single decorator

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**